api/api.go

```diff
@@ -429,12 +441,30 @@ func (api *API) listSilences(w http.ResponseWriter, r *http.Request) {
 			}, nil)
 			return
 		}
+
+		if !matchesFilterLabels(s, matchers) {
+			continue
+		}
 		sils = append(sils, s)
 	}
 
 	respond(w, sils)
 }
 
+func matchesFilterLabels(s *types.Silence, matchers []*labels.Matcher) bool {
+	sms := map[string]string{}
+	for _, m := range s.Matchers {
```

Comment:
This whole loop is less than ideal.

I'm not sure how `Matchers` is used elsewhere in the code or if it is a slice for a reason, but changing the initial parsing of the protobuf silence so that `silence.Matchers` a `map[string]*Matcher` would simplify this.

Or, if you don't care about this being in the API, that's fine too.